src/azure-cli-core/azure/cli/core/_profile.py

```diff
@@ -95,9 +95,61 @@ class CredentialType(Enum):  # pylint: disable=too-few-public-methods
     rbac = CLOUD.endpoints.active_directory_graph_resource_id
 
 
-class Profile(object):
-    def __init__(self, storage=None, auth_ctx_factory=None):
+class ReadOnlyProfile(object):
+
+    def __init__(self, storage=None):
         self._storage = storage or ACCOUNT
+
+    def get_subscription(self, subscription=None):  # take id or name
+        subscriptions = self.load_cached_subscriptions()
+        if not subscriptions:
+            raise CLIError("Please run 'az login' to setup account.")
+
+        result = [x for x in subscriptions if (
+            not subscription and x.get(_IS_DEFAULT_SUBSCRIPTION) or
+            subscription and subscription.lower() in [x[_SUBSCRIPTION_ID].lower(), x[
+                _SUBSCRIPTION_NAME].lower()])]  # pylint: disable=line-too-long
```

Comment:
And we probably should break this huge logic expression into smaller pieces for readability. The order of operations here is way too much.